### ISSUE TITLE
DuplicateMethodCall gives false positive for variables in lamdas

### ISSUE THREAD

**Issue by oliverklee** [troessner/reek#1231]
With this class …

```Ruby
class CurrencyInput < SimpleForm::Inputs::CollectionInput
  def input(wrapper_options = nil)
    input_options = merge_wrapper_options(input_html_options, wrapper_options)

    collection = Money::Currency.table
    label_method = lambda do |currency|
      "#{currency[1].fetch(:name)} (#{currency[1].fetch(:iso_code)})"
    end
    value_method = ->(currency) { currency[1].fetch(:iso_code) }

    @builder.collection_select(
      attribute_name, collection, value_method, label_method,
      input_options, input_options
    )
  end
end
```

Reek complains about this.

```
  [20, 20, 22]:DuplicateMethodCall: CurrencyInput#input calls 'currency[1]' 3 times [https://github.com/troessner/reek/blob/master/docs/Duplicate-Method-Call.md]
```

However, as these are lambdas, there are not 3 calls, but 2 + 1.

**Comment by oliverklee** [comment]
And this warning is the same problem:

  [20, 22]:DuplicateMethodCall: CurrencyInput#input calls 'currency[1].fetch(:iso_code)' 2 times [https://github.com/troessner/reek/blob/master/docs/Duplicate-Method-Call.md]

**Comment by troessner** [comment]
Oh yeah, let's fix this. @mvz do you want to tackle this? If not I can also take a stab at it after the release of Reek 5.

**Comment by mvz** [comment]
@troessner this is what #1253 was for. It turned out to be quite difficult to fix. I'm actually more and more inclined to *not* fix this or #1131, because there are many more cases where you're calling the same method on what technically is a different object, e.g.:

```
foo = make_me_a_foo
foo.bla
foo = make_me_another_foo
foo.bla
```

We can't win, really.

**Comment by troessner** [comment]
I understand that your example is kind of not fixable for us, but what about the special case with the lambdas - would that be easier?

**Comment by mvz** [comment]
Clearer, but not a lot easier. Given our current architecture, you would need to make each block its own context.

**Comment by mvz** [comment]
... and even then, if the call is not on one of the block parameters, you have to treat the whole method as one context.

**Comment by troessner** [comment]
Ok, i see. What do you think about making this a prominent part of our DuplicateMethodCall documentation?

**Comment by mvz** [comment]
That sounds like an excellent plan.

**Comment by troessner** [comment]
Closing this since we decided that we can't fix this properly and with #1350 closed to being merged.

**Comment by Jwashton** [comment]
Commenting just to document that with modern Ruby this can happen even without an explicit lambda, using pattern matching. Take the following example of a (woefully trivial and incomplete) parser of a stream of tokens:

```ruby
def parse(tokens)
  case tokens
  # base case
  in []
    []

  # "[7]"
  in ['[', value, ']', *tail]
    [{ type: :index, value: value.to_i }, *parse(tail)]

  # ".someval"
  in ['.', *tail]
    parse(tail)

  # "someval"
  in [value, *tail]
    [{ type: :key, value: }, *parse(tail)]

  end
end
```

This produces the following issue:

```
DuplicateMethodCall: Widget#parse calls 'parse(tail)' 3 times
```

I submit that this is not a duplicate method call because we're dealing with three different definitions of `tail`. I also suspect that this would be as difficult to fix as the lambda issue.
